middleware_grip.go

```diff
@@ -36,34 +36,40 @@ func setServiceLogger(r *http.Request, logger grip.Journaler) *http.Request {
 	return r.WithContext(context.WithValue(r.Context(), loggerKey, logger))
 }
 
-type logAnnotation struct {
-	key   string
-	value interface{}
-}
+type loggingAnnotations map[string]interface{}
 
 // AddLoggingAnnotation adds a key-value pair to be added to logging
-// messages used by the application logging information. There can be
-// only one annotation registered per-request.
-func AddLoggingAnnotation(r *http.Request, key string, data interface{}) *http.Request {
-	return r.WithContext(context.WithValue(r.Context(), loggingAnnotationsKey, &logAnnotation{key: key, value: data}))
+// messages used by the application logging information, overwriting
+// any previous value for the key.
+// This will noop if the logger hasn't already been set up.
+func AddLoggingAnnotation(r *http.Request, key string, value interface{}) {
+	annotations := getLoggingAnnotations(r.Context())
+	if annotations == nil {
+		return
+	}
+
+	annotations[key] = value
 }
 
-func setStartAtTime(r *http.Request, startAt time.Time) *http.Request {
-	return r.WithContext(context.WithValue(r.Context(), startAtKey, startAt))
+func setLoggingAnnotations(r *http.Request) *http.Request {
```

Comment:
Can this be called initializeLoggingAnnotations since it's just creating the empty map and not setting any values? Alternatively maybe AddLoggingAnnotation can just initialize it if it's nil.

Reply:
This function seems to me to jive with other functions that serve a similar purpose like setServiceLogger, setRequestID, setStartAtTime, and setUserForRequest. The idea of each of these functions is to set the indicated key in the context tree.

I made the function take the map as an argument. I think it makes it clearer the idea is to set the loggingAnnotationsKey in the context to whatever is passed in. Do you think that makes the function's purpose clearer?